src/Database/LSMTree/Internal/Run/Index/Compact.hs

```diff
@@ -1,21 +1,506 @@
+  only occurs with a low probability.
+
+  === Internals
+
+  TODO: explain range-finder bits, primary bits, secondary bits
+
+  @
+      Range-finder bits              Primary bits                  Secondary bits
+    |-------------------|---------------------------------------|------------------- .... {- elided -}
+     0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 .... {- elided -}
+  @
+-}
+
+-- | A compact fence-pointer index for uniformly distributed keys.
+--
+-- Compact indexes save space by storing bit-prefixes of keys.
+--
+data CompactIndex k = CompactIndex {
+    -- | A vector for finding the range of 'ciPrimary' to search based on the
+    -- range-finder bits.
+    --
+    -- TODO: elaborate.
+    ciRangeFinder          :: !(VU.Vector Word32)
+    -- | Determines the size of 'ciRangeFinder' as @2 ^ 'ciRangeFinderPrecision'
+    -- + 1@.
+  , ciRangeFinderPrecision :: !Int
+
+    -- | Maps a page @i@ to the 32-bit slice of primary bits of its minimum key.
+  , ciPrimary              :: !(VU.Vector Word32)
+
+    -- | A clash on page @i@ means that the minimum key of page @i@ and the
+    -- maximum key of the predecessor page @i-1@ have the exact same @32@
+    -- primary bits. Use the tie-breaker index to break ties.
+    --
+    -- TODO: we could use a ByteArray here instead
+  , ciClashes              :: !(VU.Vector Bit)
+    -- | Maps a page @i@ to its full minimum key, but only if there is a clash
+    -- on page @i@.
+    --
+    -- INVARIANT: a key is included in this 'IntMap' if and only if 'ciClashes'
+    -- recorded that there is a class. In code: @isJust (IntMap.lookup ix
+    -- ciTiebreaker) == ciClashes VU.! ix@.
+  , ciTieBreaker           :: !(IntMap k)
+  }
+
+{-------------------------------------------------------------------------------
+  Invariants
+-------------------------------------------------------------------------------}
+
+-- | Inclusive bounds for range-finder bit-precision.
+rangeFinderPrecisionBounds :: (Int, Int)
+rangeFinderPrecisionBounds = (0, 16)
+
+-- | https://en.wikipedia.org/wiki/Birthday_problem#Probability_of_a_shared_birthday_(collision)
+suggestRangeFinderPrecision :: Int -> Int
+suggestRangeFinderPrecision maxPages =
+    clamp $ ceiling @Double $ 2 * logBase 2 (fromIntegral maxPages)
+  where
+    (lb, ub)            = rangeFinderPrecisionBounds
+    clamp x | x < lb    = lb
+            | x > ub    = ub
+            | otherwise = x
+
+{-------------------------------------------------------------------------------
+  Queries
+-------------------------------------------------------------------------------}
+
+-- | Given a search key, find the number of the disk page that /might/ contain
+-- this key.
+--
+-- Note: there is no guarantee that the disk page contains the key. However, it
+-- is guaranteed that /no other/ disk page contain the key.
+--
+search ::
+     forall k. (SliceBits k, Integral k)
```

Comment:
Hmmm. I don't think we want `Integral k`. Consider our bytestring keys. We can extract the top 48 bits from this, but it doesn't make a lot of sense to be able to do arithmetic operations on the bytestring key type.

I think what you want is that the topBits should extract into a different type from the input type. For example the input might be ShortByteString but the result of topBits could be Word64. Then it makes sense to do arithmetic with the Word64, and there's no need to be able to do arithmetic with the key type directly.

I _think_ for the key type we just want SliceBits and Ord.

In the end, arn't we going to use the compact index at one type only? i.e. we would convert the user's key type into the binary key string and do the search with that. If so, we should at least provide SPECIALISE pragmas for the type we intend to use it at.